src/lib/migrations.ts

```diff
@@ -0,0 +1,171 @@
+    db.prepare(
+      `CREATE TABLE "_prisma_migrations" (
+        "id"                    TEXT PRIMARY KEY NOT NULL,
+        "checksum"              TEXT NOT NULL,
+        "finished_at"           DATETIME,
+        "migration_name"        TEXT NOT NULL,
+        "logs"                  TEXT,
+        "rolled_back_at"        DATETIME,
+        "started_at"            DATETIME NOT NULL DEFAULT (strftime('%Y-%m-%d %H:%M:%f', current_timestamp)),
+        "applied_steps_count"   INTEGER UNSIGNED NOT NULL DEFAULT 0
+      );`
+    ).run()
+  }
+
+  const mostRecentMigrationName: string | undefined = db
+    // Note that `finished_at` must have millisecond granularity in order for the ORDER BY to work as expected
+    // Otherwise, it won't work in the case where multiple migrations are applied in one go
+    .prepare(
+      `SELECT migration_name as name FROM '_prisma_migrations'
+      WHERE finished_at IS NOT NULL AND rolled_back_at IS NULL
+      ORDER BY finished_at DESC LIMIT 1;`
+    )
+    .get()?.name
+
+  const migrationsToApply = getUnappliedMigrations(
+    dataDir,
+    mostRecentMigrationName
+  )
+
+  migrationsToApply.forEach((migration) => {
+    const migrationFile = fs.readFileSync(migration.path, 'utf8')
+
+    const migrationId = uuidv4()
+
+    db.prepare<{
+      id: string
+      checksum: string
+      migration_name: string
+    }>(
+      "INSERT INTO '_prisma_migrations' (id, checksum, migration_name) VALUES (:id, :checksum, :migration_name);"
+    ).run({
+      id: migrationId,
+      checksum: crypto
+        .createHash('sha256')
+        .update(migrationFile.toString())
+        .digest('hex'),
+      migration_name: migration.name,
+    })
+
+    const executeMigration = db.transaction(() => {
+      db.exec(migrationFile)
+    })
+
+    try {
+      executeMigration()
+
+      // Reference for SQLite date functions: https://www.sqlite.org/lang_datefunc.html
+      db.prepare<{
+        id: string
+        finished_at: number
+      }>(
+        `UPDATE '_prisma_migrations' 
+        SET
+          finished_at = strftime('%Y-%m-%d %H:%M:%f', :finished_at / 1000, 'unixepoch'),
+          applied_steps_count = 1
+        WHERE id = :id;`
+      ).run({
+        id: migrationId,
+        finished_at: Date.now().valueOf(),
+      })
+    } catch (err) {
+      if (err instanceof SqliteError) {
+        db.prepare<{
+          id: string
+          logs: string
+          rolled_back_at: number
+        }>(
+          `UPDATE '_prisma_migrations' 
+          SET
+            logs = :logs,
+            rolled_back_at = strftime('%Y-%m-%d %H:%M:%f', :rolled_back_at / 1000, 'unixepoch')
+          WHERE id = :id;`
+        ).run({
+          id: migrationId,
+          logs: err.message,
+          rolled_back_at: Date.now().valueOf(),
+        })
+      }
+
+      throw err
+    }
+  })
+}
+
+function getUnappliedMigrations(
```

Comment:
definitely need some feedback on the approach here. basically traverses a `migrations/` directory and returns a list of migration names whose directories were created after the specified migration's corresponding directory

e.g. given a file structure where order is determined by the creation timestamp:

```
migration_first/
migration_second/
migration_third/
```

running this function with an argument of `migration_first`, for example, will return `['migration_second', 'migration_third'`

Reply:
Yes this will be fragile. On mobile (and possibly desktop) we extract this folder from the app distribution. Aren't the directory names themselves time stamps so we can just sort by name?

Reply:
from when i last checked, it wasn't but didn't spend much time trying to figure it out

Reply:
The current one is: `migrations/20220211184958_init`
I think it's core to how their migrations work (e.g. the folder name gives the ordering they should be applied).

Reply:
finally found the relevant prisma code: https://github.com/prisma/prisma-engines/blob/main/migration-engine/connectors/migration-connector/src/migrations_directory.rs#L31-L35

yeah it's a timestamp but I was expecting a unix timestamp. instead, it's just a formatted date. in other words, yeah we can just rely on that to determine order, which makes this much simpler